File: csharp/Pila/Pila/Encadenada.cs
public class Nodo
{
  public int item { get; set; }
  public Nodo? siguiente { get; set; }
}

public class Encadenada
{
  private int dimension;
  private int cantidad;
  private Nodo? tope;

  public Encadenada(int dim)
  {
    dimension = dim;
    cantidad = 0;
    tope = null;
  }

  public int Insertar(int item)
  {
    if (cantidad == dimension)
    {
      Console.WriteLine("Pila llena.");
      return -1;
    }
    Nodo nuevoNodo = new();
    nuevoNodo.item = item;
    nuevoNodo.siguiente = tope;
    tope = nuevoNodo;
    cantidad++;
    return tope.item;
  }

  public int Suprimir()
  {
    if (Vacia())
    {
      Console.WriteLine("Pilla vacia.");
      return -1;
    }
    Nodo eliminado = tope!;
    tope = tope!.siguiente;
    cantidad--;
    return eliminado!.item;
  }

  public void Recorrer()
  {
    Nodo actual = tope!;
    while (actual != null)
    {
      Console.Write(actual.item + " ");
      actual = actual.siguiente!;
    }
    Console.WriteLine();
  }

  public bool Vacia()
  {
    return cantidad == 0;
  }
}

File: python/03-divisiones-sucesivas/pilas/encadenada.py
class Nodo:
    def __init__(self, item, siguiente=None) -> None:
        self.__item = item
        self.__siguiente = siguiente

    def get_item(self):
        return self.__item

    def get_siguiente(self):
        return self.__siguiente


class Pila_Encadenada:
    __cantidad: int
    __cabeza: Nodo | None

    def __init__(self, cantidad=0) -> None:
        self.__cantidad = cantidad
        self.__cabeza = None

    def insertar(self, item):
        nuevo_nodo = Nodo(item, self.__cabeza)
        self.__cabeza = nuevo_nodo
        self.__cantidad += 1

    def eliminar(self):
        if self.__cabeza is None:
            raise IndexError("pila vacia.")
        temp = self.__cabeza.get_item()

        self.__cabeza = self.__cabeza.get_siguiente()
        self.__cantidad -= 1
        return temp

    def mostrar(self):
        actual = self.__cabeza

        while actual is not None:
            print(actual.get_item(), end=" ")
            actual = actual.get_siguiente()

    def vacia(self):
        return self.__cantidad == 0

    def longitud(self):
        return self.__cantidad


if __name__ == "__main__":
    pila = Pila_Encadenada(3)
    print("Agregando:")

    pila.insertar(1)
    pila.insertar(2)
    pila.insertar(3)

    pila.insertar(4)
    pila.mostrar()

    print("Eliminando:")
    print(pila.eliminar())
    print(pila.eliminar())
    print(pila.eliminar())

    print(pila.eliminar())

    pila.mostrar()
